feat(common): add params percent encoding encoder

solos opened the pull request:
Previously, if a http param's value contains reserved characters such as
";", the value is not encoded correctly. It was encoded and then replaced.
But change the encoding method may cause some problems for old projects.
So this commit adds param percent encoding encoder.

## PR Checklist

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: [11058](https://github.com/angular/angular/issues/11058)  [26979](https://github.com/angular/angular/issues/26979 )  


## What is the new behavior?

  If create a http with option percentEncoding=true, the params will be percent encoded.
     var params  = new HttpParams({percentEncoding: true}); 

## Does this PR introduce a breaking change?

- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

solos committed:
feat(common): add params percent encoding option

Previously, if a http param's value contains reserved characters such as
";", the value is not encoded correctly. It was encoded and then replaced.
But change the encoding method may cause some problems for old projects.
So this commit adds an option for http params percent encoding.

petebacondarwin commented on packages/common/http/src/params.ts:
```diff
@@ -125,6 +174,7 @@ export interface HttpParamsOptions {
 
   /** Encoding codec used to parse and serialize the parameters. */
   encoder?: HttpParameterCodec;
+  percentEncoding?: boolean;
```
The idea is that the new encoder would be provided via the `encoder` property of this `options` object.
So there is no need to add a new property here.

solos replied:
Yeah, adding this option is just for convenience. Without this opt, I have to import class and create a new encoder and pass it. Do you think it is unnecessary?

petebacondarwin commented on packages/common/http/src/params.ts:
```diff
@@ -143,6 +193,9 @@ export class HttpParams {
 
   constructor(options: HttpParamsOptions = {} as HttpParamsOptions) {
     this.encoder = options.encoder || new HttpUrlEncodingCodec();
+    if (!!options.percentEncoding) {
+      this.encoder = new HttpUrlPercentEncodingCodec();
+    }
```
This should be removed.

petebacondarwin commented on packages/common/http/test/params_spec.ts:
```diff
@@ -110,5 +110,16 @@ import {HttpParams} from '@angular/common/http/src/params';
+
+    describe('percent encoding toString', () => {
+      it('should encode and stringify string params', () => {
+        const body = new HttpParams({fromObject: {a: '@:$,;+=?/', b: '2'}, percentEncoding: true});
```
```suggestion
        const encoder = new HttpUrlPercentEncodingCodec();
        const body = new HttpParams({fromObject: {a: '@:$,;+=?/', b: '2'}, encoder});
```

petebacondarwin commented on packages/common/http/src/params.ts:
```diff
@@ -70,6 +70,55 @@ export class HttpUrlEncodingCodec implements HttpParameterCodec {
+ *
+ * @publicApi
+ */
+export class HttpUrlPercentEncodingCodec implements HttpParameterCodec {
```
This class needs to be exported from  [packages/common/http/public_api.ts](https://github.com/angular/angular/blob/6fe3a1de7fc452d3e1240be392333126981e3516/packages/common/http/public_api.ts#L15).

petebacondarwin replied:
And then the public API golden file will need to be updated.

petebacondarwin reviewed (changes requested):
Thanks for getting this feature moving.
I have made some inline comments.

Ideally I would also like to see some more detailed tests of more cases for this codec.

shardt68 commented:
why don't use the fix from 
https://github.com/alxhub/angular/commit/fa5a37ecd8c3aeaa1269746818e8f6b9fbf063e5
it is a general solution for the bug

petebacondarwin commented:
I was not aware of alxhub@fa5a37e. I have to admit that is a reasonable solution. Perhaps @alxhub could make a call here?

alxhub commented:
I think reviving that commit would be ideal (and fixing it - that initial work was not fully baked yet) I have two concerns about this PR:

1) the name `HttpUrlPercentEncodingCodec` is too specific to the bug. It should describe what the class does, not what it fixes.

2) as @petebacondarwin notes, the `percentEncoding` flag is unnecessary.

I'm wondering if perhaps making the fix opt-in is unnecessary. Maybe in 12 we should just fix the bug, and have an opt-out to return to the old behavior if required.

solos commented:
1. HttpUrlPercentEncodingCodec 's function is  percent encoding for params. It does not fix something. It just for  percent encoding.
2. percentEncoding is for convenience. So need not to create a new  encoder where create http params.
It has been a long time since the issue was created. I think it should be resolved in Angular 11.

solos committed:
feat(common): add params percent encoding encoder

Previously, if a http param's value contains reserved characters such as
";", the value is not encoded correctly. It was encoded and then replaced.
But change the encoding method may cause some problems for old projects.
So this commit adds an new encoder for http params percent encoding.
